onnxruntime/core/providers/coreml/model/model.mm

```diff
@@ -1,37 +1,104 @@
 
 // force the linker to create a dependency on the CoreML framework so that in MAUI usage we don't need
 // to manually do this
 asm(".linker_option \"-framework\", \"CoreML\"");
 
+using namespace onnxruntime;
+using namespace onnxruntime::coreml;
+
+namespace {
+/**
+ * Computes the static output shape used to allocate the output tensor.
+ * `inferred_shape` is the inferred shape known at model compile time. It may contain dynamic dimensions (-1).
+ * `coreml_static_shape` is the static output shape of the CoreML MLMultiArray output. It must NOT contain dynamic
+ * dimensions.
+ */
+InlinedVector<int64_t> GetStaticOutputShape(gsl::span<const int64_t> inferred_shape,
+                                            gsl::span<const int64_t> coreml_static_shape,
+                                            const logging::Logger& logger) {
+  ORT_ENFORCE(IsStaticShape(coreml_static_shape),
+              "CoreML output shape (", Shape2String(coreml_static_shape), ") is not static.");
+
+  // Special CoreML behavior notes:
+  // - Sometimes the CoreML output shape has extra leading ones.
+
+  ORT_ENFORCE(inferred_shape.size() <= coreml_static_shape.size(),
+              "CoreML static output shape (", Shape2String(coreml_static_shape),
+              ") has fewer elements than the inferred shape (", Shape2String(inferred_shape), ").");
+
+  // if coreml_static_shape has more elements, we expect them to be leading ones
+  const size_t num_leading_dimensions = coreml_static_shape.size() - inferred_shape.size();
+  const auto coreml_static_shape_common_begin = coreml_static_shape.begin() + num_leading_dimensions;
+
+  if (num_leading_dimensions > 0) {
+    const bool has_only_leading_ones =
+        std::all_of(coreml_static_shape.begin(), coreml_static_shape_common_begin,
+                    [](int64_t dim) { return dim == 1; });
+    ORT_ENFORCE(has_only_leading_ones, "CoreML static output shape (", Shape2String(coreml_static_shape),
+                ") has leading dimensions with value other than 1.");
+  }
+
+  InlinedVector<int64_t> static_shape{};
+  static_shape.reserve(inferred_shape.size());
+  std::transform(inferred_shape.begin(), inferred_shape.end(),
+                 coreml_static_shape_common_begin,
+                 std::back_inserter(static_shape),
+                 [&](int64_t inferred_dim, int64_t coreml_static_dim) {
+                   ORT_ENFORCE(inferred_dim == -1 || inferred_dim == coreml_static_dim,
+                               "CoreML static output shape (", Shape2String(coreml_static_shape),
+                               ") and inferred shape (", Shape2String(inferred_shape),
+                               ") have an inconsistent static dimensions (", coreml_static_dim, " vs. ",
+                               inferred_dim, ").");
+
+                   return inferred_dim != -1 ? inferred_dim : coreml_static_dim;
+                 });
+
+  // TODO change to VERBOSE log level
```

Comment:
update log level

Reply:
not always - I only noticed the extra leading 1's when the input shape was dynamic.

Reply:
FYI, I asked about it here. hopefully someone will respond.

https://developer.apple.com/forums/thread/734985